### PR TITLE
Add open and delivery tracking

### PR DESCRIPTION
**Pull request by jonathanlloyd**
We are about to go GA with this SDK
We should report open & delivery events because the debug console/webhooks won't work without it!

﻿- Add service worker build artefact
- Add open & delivery tracking to service worker

### CODE REVIEW

File: src/service-worker.js [hunk 1]
```diff
@@ -1,6 +1,74 @@
...
+      ? self.PusherPushNotifications.endpointOverride
+      : `https://${instanceId}.pushnotifications.pusher.com`,
+
+  _getVisibleClient: () =>
+    self.clients
+      .matchAll({
+        type: 'window',
+        includeUncontrolled: true,
+      })
+      .then(clients => clients.find(c => c.visibilityState === 'visible')),
```
Comment (TomKemp) [inline src/service-worker.js:20]: Why are we checking for visibility rather than focus?

File: src/service-worker.js [hunk 2]
```diff
@@ -1,6 +1,74 @@
...
+import doRequest from './do-request';
+import DeviceStateStore from './device-state-store';
+
 self.PusherPushNotifications = {
+  endpointOverride: null,
   onNotificationReceived: null,
+
+  _endpoint: instanceId =>
+    self.PusherPushNotifications.endpointOverride
+      ? self.PusherPushNotifications.endpointOverride
```
Comment (TomKemp) [inline src/service-worker.js:11]: How do developers set this? Should it be passed from the client?

File: src/service-worker.test.js [hunk 3]
```diff
@@ -1,6 +1,18 @@
...
+const TEST_INSTANCE_ID = 'some-instance-id';
+const TEST_PUBLISH_ID = 'some-publish-id';
+const TEST_NOTIFICATION_TITLE = 'Hi!';
+const TEST_NOTIFICATION_BODY = 'This is a test notification!';
+const TEST_NOTIFICATION_ICON = 'an-icon.png';
+
 let listeners = {};
 let shownNotifications = [];
 let openedWindows = [];
+let clients = [];
```
Comment (TomKemp) [inline src/service-worker.test.js:14]: This needs to be reset in the `beforeEach`

File: src/service-worker.js [hunk 4]
```diff
@@ -1,6 +1,74 @@
...
+      hasDisplayableContent,
+      hasData,
+    } = pusherMetadata;
+    if (!instanceId || !publishId) {
+      // Can't report this notification, fail silently.
+      return;
+    }
+
+    const deviceStateStore = new DeviceStateStore(instanceId);
+    await deviceStateStore.connect();
```
[the remark this answers is not in the document]
Comment (jonathanlloyd) [reply src/service-worker.js:40]: Yeah, this lazily creates the db

File: src/service-worker.js (hunk 1 above)
Comment (jonathanlloyd) [reply src/service-worker.js:20]: Seemed more in the spirit of the property being sent - doesn't really matter though as these properties can't be used to derive "openability" as we do for iOS/Android

File: src/service-worker.js (hunk 2 above)
Comment (jonathanlloyd) [reply src/service-worker.js:11]: Set after `importScripts` - Not intended for public use though

**Pull request merged**